### PR TITLE
prov/verbs: verbs/msg doesn't generate CQE if a context is NULL

### LINKED ISSUE #4186: fi_pingpong test hangs with verbs provider
I'm a 'newbie' to OFI and want to try running the fi_pingpong test over IB verbs. I've compiled the latest libfabrics version (1.6) and I can get it to work over sockets, but it hangs when I try the verbs provider using the msg endpoint. Any tips on getting it working?

### PR DESCRIPTION
**Pull request by dmitrygladkov**
Fixes #4186

Signed-off-by: Dmitry Gladkov <dmitry.gladkov@intel.com>

### CODE REVIEW

File: prov/verbs/src/verbs_cq.c [hunk 1]
```diff
@@ -243,7 +243,7 @@ static inline int fi_ibv_poll_outstanding_cq(struct fi_ibv_ep *ep,
 		return ret;
 
 	/* Handle WR entry when user doesn't request the completion */
-	if (!wc.wr_id) {
+	if (wc.wr_id == 0xffffffff) {
```
Comment (a-ilango) [inline prov/verbs/src/verbs_cq.c]: We should assert context != 0xffffffff in non-inject calls.

**Comment by a-ilango** [comment]
Should this be back-ported to v1.6.x?

**Comment by AddyLaddy** [comment]
I tested this patch and fi_pingpong over verbs now works for me. Thanks!

File: prov/verbs/src/verbs_cq.c (hunk 1 above)
Comment (a-ilango) [reply prov/verbs/src/verbs_cq.c]: Also, let's use a macro for this special value of wr_id.

**Comment by shefty** [comment]
The wr_id is 64-bits.  You're only setting 32-bits to 1 if this matters.

**Comment by dmitrygladkov** [comment]
> The wr_id is 64-bits. You're only setting 32-bits to 1 if this matters.

Good catch. I thought that this is 32-bits value. Fixed in the new version of the patch

File: prov/verbs/src/verbs_cq.c (hunk 1 above)
Comment (dmitrygladkov) [reply prov/verbs/src/verbs_cq.c]: - assert(...) were added.
- macro `VERBS_INJECT_FLAG ((uint64_t)-1)` were defined

**Comment by dmitrygladkov** [comment]
> Should this be back-ported to v1.6.x?

Yes, I'll prepare a patch for v1.6.x

File: prov/verbs/src/fi_verbs.h [hunk 2]
```diff
@@ -591,7 +593,7 @@ fi_ibv_process_wc(struct fi_ibv_cq *cq, struct ibv_wc *wc)
 	int ret = 1;
 
 	/* Handle WR entry when user doesn't request the completion */
-	if (!wc->wr_id)
+	if (wc->wr_id == (uint64_t)-1)
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.h]: Use VERBS_INJECT_FLAG

File: prov/verbs/src/verbs_ep.c [hunk 3]
```diff
@@ -889,6 +891,8 @@ fi_ibv_send_buf_inline(struct fi_ibv_ep *ep, struct ibv_send_wr *wr,
 {
 	struct ibv_sge sge = fi_ibv_init_sge_inline(buf, len);
 
+	assert(wr->wr_id == VERBS_INJECT_FLAG);
```
Comment (shefty) [inline prov/verbs/src/verbs_ep.c]: I think you meant !=

Comment (dmitrygladkov) [reply prov/verbs/src/verbs_ep.c]: this function is used for inject send/write. here this assert() checks that flag is set

File: prov/verbs/src/fi_verbs.h (hunk 2 above)
Comment (dmitrygladkov) [reply prov/verbs/src/fi_verbs.h]: 👍

**Pull request merged**